File: server/src/notifiers/pushover.ts
import axios from "axios";
import { AlertPayload } from "../types";
import { INotifier } from "./types";

const PUSHOVER_API = "https://api.pushover.net/1/messages.json";

const PRIORITIES: Record<string, number> = {
  DOWN: 1,
  DEGRADED: 0,
  CONTENT_CHANGED: 0,
  RECOVERED: -1,
  CVE_NEW: 0,
};

const pushover: INotifier = {
  type: "pushover",
  displayName: "Pushover",
  configSchema: {
    appToken: {
      label: "App Token",
      type: "password",
      required: true,
      placeholder: "Your Pushover application token",
    },
    userKey: {
      label: "User Key",
      type: "password",
      required: true,
      placeholder: "Your Pushover user key",
    },
  },
  async send(config: Record<string, unknown>, payload: AlertPayload): Promise<void> {
    const appToken = config.appToken as string;
    const userKey = config.userKey as string;
    if (!appToken || !userKey) throw new Error("Pushover appToken and userKey are required");

    const priority = PRIORITIES[payload.alertType] ?? 0;
    let message = payload.message;
    if (payload.diffViewUrl) {
      message += `\nDiff: ${payload.diffViewUrl}`;
    }

    const params = new URLSearchParams({
      token: appToken,
      user: userKey,
      title: `[${payload.alertType}] ${payload.serverName}`,
      message,
      priority: String(priority),
      ...(priority === 2 ? { retry: "60", expire: "3600" } : {}),
    });

    await axios.post(PUSHOVER_API, params.toString(), {
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
    });
  },
};

export { pushover };

File: vigil/src/notifiers/pushover.rs
use std::collections::HashMap;

use serde_json::Value;

use crate::types::AlertPayload;

const PUSHOVER_API: &str = "https://api.pushover.net/1/messages.json";

pub async fn send(
    config: &HashMap<String, Value>,
    payload: &AlertPayload,
) -> anyhow::Result<()> {
    let app_token = config
        .get("appToken")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("Pushover appToken is required"))?;
    let user_key = config
        .get("userKey")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("Pushover userKey is required"))?;

    let priority: i32 = match payload.alert_type.to_string().as_str() {
        "DOWN" => 1,
        "RECOVERED" => -1,
        _ => 0,
    };

    let mut message = payload.message.clone();
    if let Some(url) = &payload.diff_view_url {
        message.push_str(&format!("\nDiff: {}", url));
    }

    let mut params: Vec<(&str, String)> = vec![
        ("token", app_token.to_string()),
        ("user", user_key.to_string()),
        ("title", format!("[{}] {}", payload.alert_type, payload.server_name)),
        ("message", message),
        ("priority", priority.to_string()),
    ];

    if priority == 2 {
        params.push(("retry", "60".to_string()));
        params.push(("expire", "3600".to_string()));
    }

    reqwest::Client::new()
        .post(PUSHOVER_API)
        .form(&params)
        .send()
        .await?
        .error_for_status()?;

    Ok(())
}
